### PR TITLE
Add fine-grained check of allowed reservation states

### PR DESCRIPTION
**Pull request by danjacob-anders**
A new method Reservation.is_new_state_allowed() checks that a) the user has permission to change the state for this instance and b) that the new state is permitted, depending on current state and other factors.

Refs TTVA-102

### CODE REVIEW

File: resources/models/reservation.py [hunk 1]
```diff
@@ -357,6 +357,30 @@ def can_view_access_code(self, user):
...
         return self.resource.can_view_reservation_access_code(user)
 
+    def is_new_state_allowed(self, new_state, user):
+        """Checks that new state is permitted for this user."""
+
+        # if same state, always OK
+        if new_state == self.state:
+            return True
+
+        if not self.resource.can_approve_reservations(user):
```
Comment (jopesy) [inline resources/models/reservation.py:367]: This is interesting. I thought (unpaid) reservations cancelled by the end user were cancelled with a `PUT` request too but apparently Varaamo sends a `DELETE` request instead.

**Review by jopesy — APPROVED** [review]
Looks very good! Good job with the tests too.

File: resources/models/reservation.py (hunk 1 above)
Comment (danjacob-anders) [reply resources/models/reservation.py:367]: Yes was thinking whether `CANCELLED` needs to be here, as it's handled by another endpoint/set of permissions.

Comment (jopesy) [reply resources/models/reservation.py:367]: Let's go with this. We might need to revisit this too when cancellations for paid reservations are implemented at some point.

**Pull request merged**